src/DotNetty.Common/Utilities/HashedWheelTimer.cs

```diff
@@ -0,0 +1,712 @@
+                        Contract.Assert(sleepTimeMs <= int.MaxValue);
+                        delay = Task.Delay((int)sleepTimeMs, this.owner.CancellationToken);
+                        delay.Wait();
+                    }
+                    catch (AggregateException) when (delay != null && delay.IsCanceled)
+                    {
+                        if (Volatile.Read(ref this.owner.workerStateVolatile) == WorkerStateShutdown)
+                        {
+                            return TimeSpan.FromTicks(long.MinValue);
+                        }
+                    }
+                }
+            }
+
+            internal ISet<ITimeout> UnprocessedTimeouts() => this.unprocessedTimeouts;
+        }
+
+        sealed class HashedWheelTimeout : ITimeout
+        {
+            const int StInit = 0;
+            internal const int StCancelled = 1;
+            const int StExpired = 2;
+
+            internal readonly HashedWheelTimer timer;
+            internal readonly TimeSpan Deadline;
+
+            volatile int state = StInit;
+
+            // remainingRounds will be calculated and set by Worker.transferTimeoutsToBuckets() before the
+            // HashedWheelTimeout will be added to the correct HashedWheelBucket.
+            internal long RemainingRounds;
+
+            // This will be used to chain timeouts in HashedWheelTimerBucket via a double-linked-list.
+            // As only the workerThread will act on it there is no need for synchronization / volatile.
+            internal HashedWheelTimeout Next;
+
+            internal HashedWheelTimeout Prev;
+
+            // The bucket to which the timeout was added
+            internal HashedWheelBucket Bucket;
+
+            internal HashedWheelTimeout(HashedWheelTimer timer, ITimerTask task, TimeSpan deadline)
+            {
+                this.timer = timer;
+                this.Task = task;
+                this.Deadline = deadline;
+            }
+
+            public ITimer Timer => this.timer;
+
+            public ITimerTask Task { get; }
+
+            public bool Cancel()
+            {
+                // only update the state it will be removed from HashedWheelBucket on next tick.
+                if (!this.CompareAndSetState(StInit, StCancelled))
+                {
+                    return false;
+                }
+                // If a task should be canceled we put this to another queue which will be processed on each tick.
+                // So this means that we will have a GC latency of max. 1 tick duration which is good enough. This way
+                // we can make again use of our MpscLinkedQueue and so minimize the locking / overhead as much as possible.
+                this.timer.cancelledTimeouts.TryEnqueue(this);
+                return true;
+            }
+
+            internal void Remove()
+            {
+                HashedWheelBucket bucket = this.Bucket;
+                if (bucket != null)
```

Comment:
I'll add a comment

---
In reply to: [114477290](https://github.com/Azure/DotNetty/pull/242#discussion_r114477290) [](ancestors = 114477290)